MonIOWebApp/ClientApp/src/app/budget/budget-data.service.ts

```diff
@@ -1,30 +1,120 @@
 import { Injectable } from '@angular/core';
 import { MockApiDataService } from 'src/app/mock-services/mock-api-data.service';
 import { BehaviorSubject } from 'rxjs/internal/BehaviorSubject';
-import { Budget} from 'src/app/models/Budget';
+import { Budget, BUDGETEXPENSEENUMS} from 'src/app/models/Budget';
 import { Observable } from 'rxjs/internal/Observable';
+import { BudgetExpense } from 'src/app/models/BudgetExpense';
+import { modelGroupProvider } from '@angular/forms/src/directives/ng_model_group';
 @Injectable({
   providedIn: 'root'
 })
 export class BudgetDataService {
   budgetData: Budget;
   _budgetData: BehaviorSubject<Budget>;
   constructor(private apiData : MockApiDataService) { 
+    this._budgetData = new BehaviorSubject<Budget>(new Budget);
     this.apiData.budgetViewData.subscribe(data => {
       console.log(data);
       if(data !== null || data !== undefined){
         this.budgetData = data;
-        this._budgetData = new BehaviorSubject<Budget>(data);
+        this._budgetData.next(data);
       }else{
         this.budgetData = new Budget;
-        this._budgetData = new BehaviorSubject<Budget>(new Budget);
+        this._budgetData.next(null);
       }
-      
     });
   }
 
   getAllBudgetData():Observable<Budget>{
     return this._budgetData.asObservable();
   }
 
+
+  modifyAnExpense(expense:BudgetExpense, expenseType:string, modificationType:string){
+    
+    switch(expenseType){
+      case BUDGETEXPENSEENUMS.PreTaxExpenses: {
+        this.modifyAPreTaxExpense(expense, modificationType);
+      }
+      case BUDGETEXPENSEENUMS.RegularExpenses: {
+        this.modifyARegularExpense(expense, modificationType);
+      }
+      case BUDGETEXPENSEENUMS.SpendingCategory: {
+        this.modifyASpendingCategory(expense, modificationType);
+      }
+      case BUDGETEXPENSEENUMS.TaxExpenses: {
+        this.modifyATaxExpense(expense, modificationType);
+      }
+    }
+    this.apiData.setBudgetData(this.budgetData);
+  }
+
+  modifyAPreTaxExpense(expense:BudgetExpense, modificationType:string){
+    switch(modificationType){
+      case "ADD":{
+        this.budgetData.preTaxExpenses.concat(expense);
+      }
+      case "EDIT":{
+        var indexToEdit = this.budgetData.preTaxExpenses.map(x=>x.id).indexOf(expense.id);
+        this.budgetData.preTaxExpenses[indexToEdit] = expense;
+      }
+      case "DELETE":{
+        let filteredPreTaxExpenses = this.budgetData.preTaxExpenses.filter(x=>x.id!=expense.id);
+        this.budgetData.preTaxExpenses = this.budgetData.preTaxExpenses = filteredPreTaxExpenses;
```

Comment:
I'm confused why we need this double assignment

Reply:
no clue. Fixing.